onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp

```diff
@@ -509,68 +602,175 @@
+        MlasQNBitZeroPointsForBlksSizeInBytes<BlkBitWidth>(BlockStrideQuantB);
 
-            for (size_t nn = 0; nn < nnlen; ++nn) {
-                for (size_t k = 0, k_blk_idx = 0; k < CountK; k += BlkLen, k_blk_idx += 1) {
-                    const size_t kklen = std::min(CountK - k, BlkLen);
+    //
+    // Proceed down 16 column-wide regions of B. Dequantize and write output 16 x 16 elements at a time.
+    //
 
-                    const std::byte* b_data =
-                        QuantBDataCol + k_blk_idx * MlasQNBitBlkDataSizeInBytes(BlkBitWidth, BlkLen);
-                    const float b_s = QuantBScaleCol[k_blk_idx];
-                    const uint8_t b_z =
-                        (QuantBZeroPointCol != nullptr)
-                            ? ((k_blk_idx & 1) == 1)
-                                  ? std::to_integer<uint8_t>(QuantBZeroPointCol[k_blk_idx / 2] >> 4)
-                                  : std::to_integer<uint8_t>(QuantBZeroPointCol[k_blk_idx / 2] & std::byte{0x0F})
-                            : 8;
+    // scales of blocks from 16 adjacent columns
+    float scale[16];
+    // float conversion offsets (including zero point) of blocks from 16 adjacent columns
+    [[maybe_unused]] float offset[16];  // only used if HasZeroPoint is true
 
-                    for (size_t kk = 0; kk < kklen; ++kk) {
-                        const size_t packed_idx = kk % SubBlkLen;
+    size_t n_cols_remaining = CountN;
+    while (n_cols_remaining > 15) {
+        for (size_t k = 0, k_blk_idx = 0; k < CountK; k += BlkLen, ++k_blk_idx) {
+            for (size_t nn = 0; nn < 16; ++nn) {
+                scale[nn] = QuantBScaleCol[nn * BlockStrideQuantB + k_blk_idx];
 
-                        const bool is_low_half = packed_idx < (SubBlkLen / 2);
-                        const size_t packed_byte_idx = packed_idx % (SubBlkLen / 2);
-                        const size_t packed_range_offset = (kk / SubBlkLen) * (SubBlkLen / 2);
+                if constexpr (HasZeroPoint) {
+                    const std::byte zp_packed =
+                        QuantBZeroPointCol[nn * StrideQuantBZeroPoint + k_blk_idx / 2];
+                    const std::byte zp = ((k_blk_idx & 1) == 1)
+                                             ? (zp_packed >> 4)
+                                             : (zp_packed & std::byte{0x0F});
+                    offset[nn] = fp32_conversion::offset + std::to_integer<uint8_t>(zp);
+                }
+            }
 
-                        const std::byte b_packed = b_data[packed_range_offset + packed_byte_idx];
-                        const std::byte b_byte = is_low_half ? (b_packed & std::byte{0x0F}) : (b_packed >> 4);
-                        const float b_value = (std::to_integer<int8_t>(b_byte) - b_z) * b_s;
+            const size_t kklen = std::min(CountK - k, BlkLen);
 
-                        Dst[(k + kk) * 16 + nn] = b_value;
+            for (size_t kk = 0; kk < kklen; kk += 16) {
```

Comment:
It is better to handle kk in inner loop for better cache.

Reply:
hm, I tried it [here](https://github.com/microsoft/onnxruntime/pull/20319/commits/0dccb15f35f11ba66b14291b4d794a1e939937b1) but the performance didn't improve.